pkg/sql/sqlstats/sslocal/app_stats.go

```diff
@@ -0,0 +1,430 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package sslocal
+
+import (
+	"context"
+	"sync/atomic"
+	"unsafe"
+
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/sql/execstats"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlstats"
+	"github.com/cockroachdb/cockroach/pkg/util/mon"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+)
+
+// TODO(arul): The fields on stmtKey should really be immutable fields on
+// stmtStats which are set once (on first addition to the map). Instead, we
+// should use stmtID (which is a hashed string of the fields below) as the
+// stmtKey.
+type stmtKey struct {
+	anonymizedStmt string
+	failed         bool
+	implicitTxn    bool
+	database       string
+}
+
+func (s stmtKey) String() string {
+	if s.failed {
+		return "!" + s.anonymizedStmt
+	}
+	return s.anonymizedStmt
+}
+
+func (s stmtKey) size() int64 {
+	return int64(unsafe.Sizeof(s)) + int64(len(s.anonymizedStmt)) + int64(len(s.database))
+}
+
+const invalidStmtID = 0
+
+// appStats holds per-application statistics.
```

Comment:
One thing that's interesting about this is that it doesn't really have much of anything to do with an app. It just happens to be the case that we bucket by app. Maybe something like `ssmemstorage`? I think abstracting this a bit might let you make the telemetry stuff feel less like a hack. 

Either way, any cleanup in that direction should follow in a separate commit and maybe even a separate PR. This is already great progress.